src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,349 @@
+
+  override def sizeHint(hint: Int) = if (hint >= mask) resize(hint + 1)
+
+  override def length = end_-(start)
+
+  override def isEmpty = start == end
+
+  override def nonEmpty = start != end
+
+  override def clone() = new ArrayDeque(array.clone, start, end)
+
+  /**
+    * Note: This does not actually resize the internal representation.
+    * See clearAndShrink if you want to also resize internally
+    */
+  override def clear() = {
+    nullify()
+    start = 0
+    end = 0
+  }
+
+  /**
+    * clears this buffer and shrinks to @param size
+    *
+    * @param size
+    * @return
+    */
+  def clearAndShrink(size: Int = ArrayDeque.DefaultInitialSize): this.type = {
+    require(size >= 0, s"Non-negative size required")
+    set(array = ArrayDeque.alloc(size), start = 0, end = 0)
+    this
+  }
+
+  override def slice(from: Int, until: Int) = {
+    val left = if (from <= 0) 0 else if (from >= size) size else from
+    val right = if (until <= 0) 0 else if (until >= size) size else until
+    val len = right - left
+    if (len <= 0) {
+      ArrayDeque.empty[A]
+    } else if (len >= size) {
+      clone()
+    } else {
+      val array2 = copySliceToArray(srcStart = left, dest = ArrayDeque.alloc(len), destStart = 0, maxItems = len)
+      new ArrayDeque(array2, 0, len)
+    }
+  }
+
+  override def sliding(window: Int, step: Int) = {
+    require(window > 0 && step > 0, s"window=$window and step=$step, but both must be positive")
+    val lag = (window - step) max 0
+    Iterator.range(start = 0, end = length - lag, step = step).map(i => slice(i, i + window))
+  }
+
+  override def grouped(n: Int) = sliding(n, n)
+
+  override def copyToArray[B >: A](dest: Array[B], destStart: Int, len: Int) =
+    copySliceToArray(srcStart = 0, dest = dest, destStart = destStart, maxItems = len)
+
+  override def companion = ArrayDeque
+
+  override def result() = this
+
+  override def stringPrefix = "ArrayDeque"
+
+  override def toArray[B >: A: ClassTag] =
+    copySliceToArray(srcStart = 0, dest = new Array[B](size), destStart = 0, maxItems = size)
+
+  /**
+    * This is a more general version of copyToArray - this also accepts a srcStart unlike copyToArray
+    * This copies maxItems elements from this collections srcStart to dest's destStart
+    * If we reach the end of either collections before we could copy maxItems, we simply stop copying
+    *
+    * @param dest
+    * @param srcStart
+    * @param destStart
+    * @param maxItems
+    */
+  def copySliceToArray(srcStart: Int, dest: Array[_], destStart: Int, maxItems: Int): dest.type = {
+    ArrayDeque.checkIndex(srcStart, this)
+    ArrayDeque.checkIndex(destStart, dest)
+    val toCopy = maxItems min (size - srcStart) min (dest.length - destStart)
+    if (toCopy > 0) {
+      val startIdx = start_+(srcStart)
+      val block1 = toCopy min (array.length - startIdx)
+      Array.copy(src = array, srcPos = startIdx, dest = dest, destPos = destStart, length = block1)
```

Comment:
Are you sure? Both this and copyOfRange call system.arrayCopy. See line 3294 here: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/Arrays.java#3253